Source/Core/Core/PowerPC/MMU.cpp

```diff
@@ -1123,26 +1112,143 @@ static __forceinline u32 TranslatePageAddress(const u32 address, const XCheckTLB
+
+    if ((batu.BEPI & batu.BL) != 0)
+    {
+      // With a valid BAT, the simplest way to match is
+      // (input & ~BL_mask) == BEPI. For now, assume it's
+      // implemented this way for invalid BATs as well.
+      WARN_LOG(POWERPC, "Bad BAT setup: BEPI overlaps BL");
+      continue;
+    }
+    if ((batl.BRPN & batu.BL) != 0)
+    {
+      // With a valid BAT, the simplest way to translate is
+      // (input & BL_mask) | BRPN_address. For now, assume it's
+      // implemented this way for invalid BATs as well.
+      WARN_LOG(POWERPC, "Bad BAT setup: BPRN overlaps BL");
+    }
+    if (CountSetBits((u32)(batu.BL + 1)) != 1)
+    {
+      // With a valid BAT, the simplest way of masking is
+      // (input & ~BL_mask) for matching and (input & BL_mask) for
+      // translation. For now, assume it's implemented this way for
+      // invalid BATs as well.
+      WARN_LOG(POWERPC, "Bad BAT setup: invalid mask in BL");
+    }
+    for (u32 j = 0; j < (1 << 11); ++j)
+    {
+      // Enumerate all bit-patterns which fit within the given mask.
+      if ((j & batu.BL) == j)
+      {
+        // This bit is a little weird: if BRPN & j != 0, we end up with
+        // a strange mapping. Need to check on hardware.
+        u32 address = (batl.BRPN | j) << BAT_INDEX_SHIFT;
+
+        // The bottom bit is whether the translation is valid; the second
+        // bit from the bottom is whether we can use the fastmem arena.
+        u32 valid_bit = 0x1;
+        if (Memory::bFakeVMEM && ((address & 0xFE000000) == 0x7E000000))
+          valid_bit = 0x3;
+        else if (address < Memory::REALRAM_SIZE)
+          valid_bit = 0x3;
+        else if (Memory::m_pEXRAM && (address >> 28) == 0x1 &&
+                 (address & 0x0FFFFFFF) < Memory::EXRAM_SIZE)
+          valid_bit = 0x3;
+        else if ((address >> 28) == 0xE && (address < (0xE0000000 + Memory::L1_CACHE_SIZE)))
+          valid_bit = 0x3;
+
+        // (BEPI | j) == (BEPI & ~BL) | (j & BL).
+        bat_table[batu.BEPI | j] = address | valid_bit;
+      }
+    }
+  }
+}
+
+static void UpdateFakeMMUBat(u32* bat_table, u32 start_addr)
+{
+  for (unsigned i = 0; i < (0x10000000 >> BAT_INDEX_SHIFT); ++i)
+  {
+    // Map from 0x4XXXXXXX or 0x7XXXXXXX to the range
+    // [0x7E000000,0x80000000).
+    u32 e_address = i + (start_addr >> BAT_INDEX_SHIFT);
+    u32 p_address = 0x7E000003 | ((i << BAT_INDEX_SHIFT) & Memory::FAKEVMEM_MASK);
+    bat_table[e_address] = p_address;
+  }
+}
+
+void DBATUpdated()
+{
+  memset(dbat_table, 0, sizeof(dbat_table));
+  UpdateBATs(dbat_table, SPR_DBAT0U);
+  bool extended_bats = SConfig::GetInstance().bWii && HID4.SBE;
+  if (extended_bats)
+    UpdateBATs(dbat_table, SPR_DBAT4U);
+  if (Memory::bFakeVMEM)
+  {
+    // In Fake-MMU mode, insert some extra entries into the BAT tables.
+    UpdateFakeMMUBat(dbat_table, 0x40000000);
+    UpdateFakeMMUBat(dbat_table, 0x70000000);
+  }
+  Memory::UpdateLogicalMemory(dbat_table);
+  JitInterface::ClearSafe();
```

Comment:
Certain load/store optimizations depend on the BAT state; off the top of my head, IsOptimizableRAMAddress, IsOptimizableMMIOAddress, IsOptimizableGatherPipeAddress.  None of those are really critical, but there isn't really any reason to get rid of them.